### ISSUE TITLE
Regression likely contains a bug

### ISSUE THREAD

**Issue by rikhuijzer** [rikhuijzer/SIRUS.jl#27]
There seems to be something wrong with the regression-case since the performance decreases when the number of `max_rules` increases. This should be the other way around. It could be that this is related to the weight calculation because the weight calculation has not much effect on the classification performance. 

Maybe the weights should take into account the rule frequencies? Currently, this is sort of used in the weight calculation via a rule space where each rule is a binary feature indicating whether the datapoint satisfies the rule. However, if I understand correctly, this is a shortcut compared to using the rules and rule frequencies as established during the tree fitting procedure.

**Comment by gdalle** [comment]
Seems like an important issue to fix for publication?

https://github.com/openjournals/joss-reviews/issues/5786

**Comment by rikhuijzer** [comment]
> Seems like an important issue to fix for publication?

@gdalle yes it would be very nice if this could be fixed. You are very much right on this. I now ran the comparison against the original implementation in #58 and it looks like the original algorithm performs similar on the classification and boston (regression) datasets, so that is fine. Alas, the original algorithm performs much better on the `make_regression` dataset.

I have already spent a few weeks on trying to improve the regression implementation, see the long list of PRs linking to https://github.com/rikhuijzer/SIRUS.jl/issues/13 and especially https://github.com/rikhuijzer/SIRUS.jl/pull/22 for details. Unfortunately, I don't have time to fix this in the coming months since I'm busy finishing my PhD. I find bug hunting in algorithms enjoyable but especially time consuming.

Regardless, I still think that the paper adds value to the scientific field even without having "fixed" this issue. The Julia implementation is much easier to understand, which hopefully allows other people to help in finding the bug or to build even better algorithms on top of it. Also the Julia version has implemented multi-class classification whereas the R version appears to have not.

**Comment by gdalle** [comment]
I understand and I agree with your stance. Not closing this but will indicate on the review that it is non blocking, as long as the performance gap is listed in the paper among future perspectives.
Nice work on the RCall wrap!

**Comment by rikhuijzer** [comment]
> I understand and I agree with your stance. Not closing this but will indicate on the review that it is non blocking, as long as the performance gap is listed in the paper among future perspectives.

Thank you, Guillaume, for seeing it as non blocking as long as we list the performance gap (which I believe we did, but correct me if I'm wrong).

> Nice work on the RCall wrap!

Haha yes very nice of you to notice that! I expected that it would be a big mess but it turned out to be surprisingly easy. Credits to the RCall.jl people!

**Comment by rikhuijzer** [comment]
A quick note to self before I forget. I think I got it (https://github.com/rikhuijzer/SIRUS.jl/pull/71). There are two problems:

1. The base forest model is not as accurate as it could be. The `StableForestRegressor` performs poorer than the `DecisionTreeRegressor` on the `boston` and `make_regression` datasets. In turn, this will also lead to lower `StableRulesRegressor`.
2. Just like linear models, rule-based model is very sensitive to the choice of `lambda` hyperparameter (because under the hood `lambda` goes into a linear model to choose the weights). With a poor `lambda`, strange things can happen such as performance _decreasing_ when _increasing_ the number of rules. This effect can (sometimes?) be fixed by choosing a more suitable `lambda`.
